tests/storage/glusterfs/conftest.py

```diff
@@ -0,0 +1,128 @@
+
+@pytest.fixture(scope='session')
+def pool_with_glusterfs(host):
+    for h in host.pool.hosts:
+        _setup_host_with_glusterfs(h)
+
+    yield
+    for h in host.pool.hosts:
+        _teardown_host_with_glusterfs(h)
+
+@pytest.fixture(scope='session')
+def gluster_disk(host, sr_disk):
+    print(">> Format sr_disk %s and mount it" % sr_disk)
+    device = '/dev/' + sr_disk
+    host.ssh(['mkfs.ext4', '-F', device])
+    host.ssh(['mkdir', '-p', '/mnt/sr_disk'])
+    host.ssh(['cp', '-f', '/etc/fstab', '/etc/fstab.orig'])
+    host.ssh(['echo', '/dev/sdb /mnt/sr_disk ext4 defaults 0 0', '>>/etc/fstab'])
+    host.ssh(['mount', '/mnt/sr_disk'])
+    yield
+    host.ssh(['cp', '-f', '/etc/fstab.orig', '/etc/fstab'])
+    try:
+        host.ssh(['umount', '/mnt/sr_disk'])
+        host.ssh(['rmdir', '/mnt/sr_disk'])
+    except Exception:
+        pass
+
+@pytest.fixture(scope='session')
+def gluster_volume_started(host, gluster_disk):
+    hosts = host.pool.hosts
+    assert len(hosts) > 1, "GlusterFS must be run on at least a 2 hosts pool"
+
+    host_probed = False
+    for h in hosts:
+        h.ssh(['mkdir', '-p', '/mnt/sr_disk/vol0/brick0'])
+        if h.hostname_or_ip != host.hostname_or_ip:
+            host.ssh(['gluster', 'peer', 'probe', h.hostname_or_ip])
+        elif not host_probed:
+            h.ssh(['gluster', 'peer', 'probe', host.hostname_or_ip])
+            host_probed = True # Only probe host once from one of the other hosts.
+
+    print(">> create and start gluster volume vol0")
+    gluster_cmd = ['gluster', 'volume', 'create', 'vol0', 'replica', str(len(hosts))]
+    for h in hosts:
+        gluster_cmd.append('%s:/mnt/sr_disk/vol0/brick0' % h.hostname_or_ip)
+
+    gluster_cmd.append('force')
+    host.ssh(gluster_cmd)
+
+    host.ssh(['gluster', 'volume', 'start', 'vol0'])
+    yield
+    print("<< stop and delete gluster volume vol0")
+    host.ssh(['gluster', '--mode=script', 'volume', 'stop', 'vol0'])
+    host.ssh(['gluster', '--mode=script', 'volume', 'delete', 'vol0'])
+    host.ssh(['rm', '-rf', '/mnt/sr_disk/vol0'])
+
+
+@pytest.fixture(scope='session')
+def glusterfs_device_config(host):
+    backup_servers = []
+    for h in host.pool.hosts:
+        hostname_or_ip = h.hostname_or_ip
+        if hostname_or_ip == host.hostname_or_ip:
+            continue
+        backup_servers.append(hostname_or_ip)
+
+    return {
+        'server': '%s:/vol0' % host.hostname_or_ip,
+        'backupservers': ':'.join(backup_servers)
+    }
+
+@pytest.fixture(scope='session')
+def glusterfs_sr(host, pool_with_glusterfs, gluster_volume_started, glusterfs_device_config):
+    """ A GlusterFS SR on first host. """
+    # Create the SR
+    sr = host.sr_create('glusterfs', "GlusterFS-SR", glusterfs_device_config, shared=True)
+    yield sr
+    # teardown
+    try:
+        sr.destroy()
+    except Exception as e:
+        print("SR destroy failed: %s" % e)
```

Comment:
Can you test (not on a physical pool because this would leave things dirty) whether an exception raised in this fixture's teardown would stop the other fixtures from being teared down. I think they would still be, so it would be better to let the exception be raised.

Reply:
I know for a fact that first exception raised in teardown stop the rest of the teardown
i had a lot of cleanings by hand to do why developping the tests x)